packages/grafana-data/src/vector/FunctionalVector.ts

```diff
@@ -15,32 +15,170 @@ export abstract class FunctionalVector<T = any> implements Vector<T>, Iterable<T
     }
   }
 
+  set(index: number, value: any): void {
+    throw 'unsupported operation';
+  }
+
+  add(value: T): void {
+    throw 'unsupported operation';
+  }
+
+  push(...vals: T[]): number {
+    for (const v of vals) {
+      this.add(v);
+    }
+    return this.length;
+  }
+
   // Implement "iterable protocol"
   [Symbol.iterator]() {
     return this.iterator();
   }
 
-  forEach(iterator: (row: T) => void) {
+  forEach(iterator: (row: T, index: number, array: T[]) => void): void {
     return vectorator(this).forEach(iterator);
   }
 
-  map<V>(transform: (item: T, index: number) => V) {
+  map<V>(transform: (item: T, index: number, array: T[]) => V): V[] {
     return vectorator(this).map(transform);
   }
 
-  filter(predicate: (item: T) => boolean): T[] {
+  filter(predicate: (item: T, index: number, array: T[]) => boolean): T[] {
     return vectorator(this).filter(predicate);
   }
 
+  at(index: number): T | undefined {
+    return this.get(index);
+  }
+
   toArray(): T[] {
     return vectorToArray(this);
   }
 
   toJSON(): any {
     return this.toArray();
   }
+
+  //--------------------------
+  // Method not implemented
+  //--------------------------
+
+  [n: number]: T;
+
+  pop(): T | undefined {
+    throw new Error('Method not implemented.');
+  }
+  concat(...items: Array<ConcatArray<T>>): T[];
+  concat(...items: Array<T | ConcatArray<T>>): T[] {
+    throw new Error('Method not implemented.');
+  }
+  join(separator?: string | undefined): string {
+    throw new Error('Method not implemented.');
+  }
+  reverse(): T[] {
+    throw new Error('Method not implemented.');
+  }
+  shift(): T | undefined {
+    throw new Error('Method not implemented.');
+  }
+  slice(start?: number | undefined, end?: number | undefined): T[] {
+    throw new Error('Method not implemented.');
+  }
+  sort(compareFn?: ((a: T, b: T) => number) | undefined): this {
+    throw new Error('Method not implemented.');
+  }
+  splice(start: number, deleteCount?: number | undefined): T[];
+  splice(start: number, deleteCount: number, ...items: T[]): T[] {
+    throw new Error('Method not implemented.');
+  }
+  unshift(...items: T[]): number {
+    throw new Error('Method not implemented.');
+  }
```

Comment:
This bit seems a bit icky, do we need to implement all these?

Reply:
I am torn here... the value is that you can use arrays without wrapping in `isArray()` or other typescript voodo logic -- but that said perhaps excluding the more esoteric functions would make things more clear while there is the *super super rare chance that the data is backed by the Vector interface. (after all 3 parts of this PR)

Reply:
ah, I see. Think I understand. This is just so that we can use the `values: T[] | Vector<T>`  and have both treated the same. And at runtime, we expect very rare cases of actually seeing data that uses ArrayVector as all will use arrays?  Do we switch out ArrayVectors to arrays runRequest or applyFieldOverrides?

Reply:
Just not sure I understand. If we start doing `field.values.splice(0, 1)` in a visualization and the data comes from a datasource that uses ArrayVector values won't that throw Not implemented exception?

Reply:
I'm afraid of the same thing. Seems to me like this is going to impact the actual usage of vectors if we really want to treat them as arrays. I think we "can" guarantee that in core, but I don't know about external usages that may face issues with exceptions from this.

Reply:
We can get all the read-only options to work -- it is the mutating ones that might throw exceptions.  I don't think this is too different than if upstream has an immutable proxy or `Object.freeze()`